File: scripts/update_tileset.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

/**
 * Script to update Tiled map with all objects from assets directory
 * This ensures all objects are included in the tileset with proper GIDs
 */

const ASSETS_DIR = path.join(__dirname, '../assets/objects');
const MAP_FILE = path.join(__dirname, '../assets/rooms/room_reception2.json');

// Object types to include
const OBJECT_TYPES = [
    'bag', 'bin', 'briefcase', 'laptop', 'phone', 'pc', 'note', 'notes',
    'safe', 'suitcase', 'office-misc', 'plant', 'chair', 'picture', 'table'
];

function getAllObjectFiles() {
    const files = [];
    
    try {
        const entries = fs.readdirSync(ASSETS_DIR, { withFileTypes: true });
        
        for (const entry of entries) {
            if (entry.isFile() && entry.name.endsWith('.png')) {
                const baseName = entry.name.replace('.png', '');
                const category = getObjectCategory(baseName);
                
                files.push({
                    filename: entry.name,
                    basename: baseName,
                    category: category,
                    path: `../objects/${entry.name}`
                });
            }
        }
    } catch (error) {
        console.error('Error reading assets directory:', error);
        return [];
    }
    
    return files.sort((a, b) => a.basename.localeCompare(b.basename));
}

function getObjectCategory(filename) {
    const baseName = filename.replace('.png', '');
    
    for (const type of OBJECT_TYPES) {
        if (baseName.includes(type)) {
            return type;
        }
    }
    
    return 'misc';
}

function findHighestGID(mapData) {
    let highestGID = 0;
    
    // Find the highest GID in the most recent objects tileset
    const objectsTilesets = mapData.tilesets.filter(ts => 
        ts.name === 'objects' || ts.name.includes('objects/')
    );
    
    if (objectsTilesets.length > 0) {
        const latestTileset = objectsTilesets[objectsTilesets.length - 1];
        highestGID = latestTileset.firstgid + (latestTileset.tilecount || 0);
    }
    
    return highestGID;
}

function createTilesetEntry(file, gid) {
    return {
        id: gid - 1, // Tiled uses 0-based indexing
        image: file.path,
        imageheight: 16, // Default size, will be updated by Tiled
        imagewidth: 16
    };
}

function updateMapFile(objectFiles) {
    try {
        const mapData = JSON.parse(fs.readFileSync(MAP_FILE, 'utf8'));
        
        // Find the latest objects tileset
        const objectsTilesets = mapData.tilesets.filter(ts => 
            ts.name === 'objects' || ts.name.includes('objects/')
        );
        
        if (objectsTilesets.length === 0) {
            console.error('No objects tileset found in map file');
            return;
        }
        
        const latestTileset = objectsTilesets[objectsTilesets.length - 1];
        const startGID = latestTileset.firstgid + (latestTileset.tilecount || 0);
        
        console.log(`Found latest objects tileset with firstgid: ${latestTileset.firstgid}`);
        console.log(`Current tilecount: ${latestTileset.tilecount || 0}`);
        console.log(`Next available GID: ${startGID}`);
        
        // Check which objects are missing
        const existingImages = new Set();
        if (latestTileset.tiles) {
            latestTileset.tiles.forEach(tile => {
                if (tile.image) {
                    const filename = path.basename(tile.image);
                    existingImages.add(filename);
                }
            });
        }
        
        const missingObjects = objectFiles.filter(file => 
            !existingImages.has(file.filename)
        );
        
        console.log(`Found ${objectFiles.length} total objects`);
        console.log(`Found ${existingImages.size} existing objects in tileset`);
        console.log(`Missing ${missingObjects.length} objects`);
        
        if (missingObjects.length === 0) {
            console.log('All objects are already in the tileset!');
            return;
        }
        
        // Add missing objects to tileset
        const newTiles = [];
        let currentGID = startGID;
        
        for (const file of missingObjects) {
            const tileEntry = createTilesetEntry(file, currentGID);
            newTiles.push(tileEntry);
            currentGID++;
        }
        
        // Update tileset
        if (!latestTileset.tiles) {
            latestTileset.tiles = [];
        }
        
        latestTileset.tiles.push(...newTiles);
        latestTileset.tilecount = (latestTileset.tilecount || 0) + newTiles.length;
        
        console.log(`Added ${newTiles.length} new objects to tileset`);
        console.log(`New tilecount: ${latestTileset.tilecount}`);
        
        // Write updated map file
        fs.writeFileSync(MAP_FILE, JSON.stringify(mapData, null, 2));
        console.log(`Updated map file: ${MAP_FILE}`);
        
        // Generate Tiled command to open the map
        console.log('\nNext steps:');
        console.log('1. Open the map in Tiled Editor');
        console.log('2. The new objects should be available in the tileset');
        console.log('3. Place objects in your layers as needed');
        console.log('4. Save the map');
        
    } catch (error) {
        console.error('Error updating map file:', error);
    }
}

function main() {
    console.log('🔧 Tileset Update Script');
    console.log('========================');
    
    // Check if assets directory exists
    if (!fs.existsSync(ASSETS_DIR)) {
        console.error(`Assets directory not found: ${ASSETS_DIR}`);
        process.exit(1);
    }
    
    // Check if map file exists
    if (!fs.existsSync(MAP_FILE)) {
        console.error(`Map file not found: ${MAP_FILE}`);
        process.exit(1);
    }
    
    console.log(`Scanning assets directory: ${ASSETS_DIR}`);
    const objectFiles = getAllObjectFiles();
    
    if (objectFiles.length === 0) {
        console.log('No object files found in assets directory');
        return;
    }
    
    console.log(`Found ${objectFiles.length} object files`);
    
    // Group by category
    const byCategory = {};
    objectFiles.forEach(file => {
        if (!byCategory[file.category]) {
            byCategory[file.category] = [];
        }
        byCategory[file.category].push(file);
    });
    
    console.log('\nObjects by category:');
    Object.entries(byCategory).forEach(([category, files]) => {
        console.log(`  ${category}: ${files.length} files`);
    });
    
    console.log('\nUpdating map file...');
    updateMapFile(objectFiles);
    
    console.log('\n✅ Script completed!');
}

if (require.main === module) {
    main();
}

module.exports = { getAllObjectFiles, updateMapFile };

File: planning_notes/npc/movement/update_tileset.py
#!/usr/bin/env python3

import os
import json
import glob
from pathlib import Path

"""
Script to update Tiled map with all objects from assets directory
This ensures all objects are included in the tileset with proper GIDs
"""

ASSETS_DIR = "assets/objects"
MAP_FILE = "assets/rooms/room_reception2.json"

# Object types to include
OBJECT_TYPES = [
    'bag', 'bin', 'briefcase', 'laptop', 'phone', 'pc', 'note', 'notes',
    'safe', 'suitcase', 'office-misc', 'plant', 'chair', 'picture', 'table'
]

def get_all_object_files():
    """Get all object files from the assets directory"""
    files = []
    
    if not os.path.exists(ASSETS_DIR):
        print(f"❌ Assets directory not found: {ASSETS_DIR}")
        return files
    
    for file_path in glob.glob(os.path.join(ASSETS_DIR, "*.png")):
        filename = os.path.basename(file_path)
        basename = filename.replace('.png', '')
        category = get_object_category(basename)
        
        files.append({
            'filename': filename,
            'basename': basename,
            'category': category,
            'path': f"../objects/{filename}"
        })
    
    return sorted(files, key=lambda x: x['basename'])

def get_object_category(filename):
    """Determine the category of an object based on its filename"""
    for obj_type in OBJECT_TYPES:
        if obj_type in filename:
            return obj_type
    return 'misc'

def find_latest_objects_tileset(map_data):
    """Find the latest objects tileset in the map data"""
    objects_tilesets = []
    
    for tileset in map_data.get('tilesets', []):
        if tileset.get('name') == 'objects' or 'objects/' in tileset.get('name', ''):
            objects_tilesets.append(tileset)
    
    if not objects_tilesets:
        return None
    
    # Return the last one (most recent)
    return objects_tilesets[-1]

def create_tileset_entry(file_info, gid):
    """Create a tileset entry for a file"""
    return {
        "id": gid - 1,  # Tiled uses 0-based indexing
        "image": file_info['path'],
        "imageheight": 16,  # Default size, will be updated by Tiled
        "imagewidth": 16
    }

def update_map_file(object_files):
    """Update the map file with missing objects"""
    try:
        with open(MAP_FILE, 'r') as f:
            map_data = json.load(f)
        
        # Find the latest objects tileset
        latest_tileset = find_latest_objects_tileset(map_data)
        
        if not latest_tileset:
            print("❌ No objects tileset found in map file")
            return False
        
        print(f"📋 Found latest objects tileset with firstgid: {latest_tileset.get('firstgid', 0)}")
        print(f"📊 Current tilecount: {latest_tileset.get('tilecount', 0)}")
        
        # Check which objects are missing
        existing_images = set()
        if 'tiles' in latest_tileset:
            for tile in latest_tileset['tiles']:
                if 'image' in tile:
                    filename = os.path.basename(tile['image'])
                    existing_images.add(filename)
        
        missing_objects = [f for f in object_files if f['filename'] not in existing_images]
        
        print(f"📁 Found {len(object_files)} total objects")
        print(f"✅ Found {len(existing_images)} existing objects in tileset")
        print(f"❌ Missing {len(missing_objects)} objects")
        
        if not missing_objects:
            print("🎉 All objects are already in the tileset!")
            return True
        
        # Add missing objects to tileset
        start_gid = latest_tileset.get('firstgid', 0) + latest_tileset.get('tilecount', 0)
        new_tiles = []
        
        for i, file_info in enumerate(missing_objects):
            gid = start_gid + i
            tile_entry = create_tileset_entry(file_info, gid)
            new_tiles.append(tile_entry)
        
        # Update tileset
        if 'tiles' not in latest_tileset:
            latest_tileset['tiles'] = []
        
        latest_tileset['tiles'].extend(new_tiles)
        latest_tileset['tilecount'] = latest_tileset.get('tilecount', 0) + len(new_tiles)
        
        print(f"➕ Added {len(new_tiles)} new objects to tileset")
        print(f"📊 New tilecount: {latest_tileset['tilecount']}")
        
        # Write updated map file
        with open(MAP_FILE, 'w') as f:
            json.dump(map_data, f, indent=2)
        
        print(f"💾 Updated map file: {MAP_FILE}")
        return True
        
    except Exception as e:
        print(f"❌ Error updating map file: {e}")
        return False

def main():
    """Main function"""
    print("🔧 Tileset Update Script")
    print("========================")
    
    # Check if map file exists
    if not os.path.exists(MAP_FILE):
        print(f"❌ Map file not found: {MAP_FILE}")
        return
    
    print(f"📂 Scanning assets directory: {ASSETS_DIR}")
    object_files = get_all_object_files()
    
    if not object_files:
        print("❌ No object files found in assets directory")
        return
    
    print(f"📁 Found {len(object_files)} object files")
    
    # Group by category
    by_category = {}
    for file_info in object_files:
        category = file_info['category']
        if category not in by_category:
            by_category[category] = []
        by_category[category].append(file_info)
    
    print("\n📊 Objects by category:")
    for category, files in by_category.items():
        print(f"  {category}: {len(files)} files")
    
    print("\n🔄 Updating map file...")
    success = update_map_file(object_files)
    
    if success:
        print("\n✅ Script completed successfully!")
        print("\n📝 Next steps:")
        print("1. Open the map in Tiled Editor")
        print("2. Check that all objects are available in the tileset")
        print("3. Place any missing objects in your layers")
        print("4. Save the map")
        print("\n🎯 This script ensures all objects from assets/objects/ are included in the tileset!")
    else:
        print("\n❌ Script failed. Please check the errors above.")

if __name__ == "__main__":
    main()
